types.go

```diff
@@ -278,16 +281,41 @@ func (source *Source) Metadata() []MetadataField {
 
 func (source *Source) AuthenticateToECR() bool {
 	logrus.Warnln("ECR integration is experimental and untested")
-	mySession := session.Must(session.NewSession(&aws.Config{
-		Region:      aws.String(source.AwsRegion),
-		Credentials: credentials.NewStaticCredentials(source.AwsAccessKeyId, source.AwsSecretAccessKey, source.AwsSessionToken),
-	}))
+
+	var mySession *session.Session = nil
+
+  if source.AwsEC2Credentials {
+		mySession = session.Must(session.NewSession(&aws.Config{
+			Region:      aws.String(source.AwsRegion),
+			Credentials: credentials.NewCredentials( &ec2rolecreds.EC2RoleProvider{Client: ec2metadata.New(session.New()),} ),
+		}))
+
+	} else {
+		mySession = session.Must(session.NewSession(&aws.Config{
+			Region:      aws.String(source.AwsRegion),
+			Credentials: credentials.NewStaticCredentials(source.AwsAccessKeyId, source.AwsSecretAccessKey, source.AwsSessionToken),
+		}))
+	}
+
+
 
 	var config aws.Config
 
 	// If a role arn has been supplied, then assume role and get a new session
-	if source.AwsRoleArn != "" {
-		config = aws.Config{Credentials: stscreds.NewCredentials(mySession, source.AwsRoleArn)}
+	if source.AwsRoleArns != "" {
+		awsRoleArnsList := strings.Split(source.AwsRoleArns, ",")
+
+		for roleIndex, roleArn := range (awsRoleArnsList) {
+			if (roleIndex < len(awsRoleArnsList)-1) {
+				mySession = session.Must(session.NewSession(&aws.Config{
+					Region:      aws.String(source.AwsRegion),
+					Credentials: stscreds.NewCredentials(mySession, roleArn),
+				}))
+			} else {
+				config = aws.Config{Credentials: stscreds.NewCredentials(mySession, roleArn)}
+			}
+		}
```

Comment:
Just iterate over a slice rather than branching inside the loop. Something like this:

```suggestion
		for roleIndex, roleArn := range (awsRoleArnsList[:len(awsRoleArnsList)-1]) {
			mySession = session.Must(session.NewSession(&aws.Config{
				Region:      aws.String(source.AwsRegion),
				Credentials: stscreds.NewCredentials(mySession, roleArn),
			}))
		}
		lastAwsRoleArn = awsRoleArnsList[len(awsRoleArnsList)-1]
		config = aws.Config{Credentials: stscreds.NewCredentials(mySession, lastAwsRoleArn)}
```

Thinking more broadly, this is probably a smell that there's a semantic distinction between the last ARN in the list and the others, which perhaps should be represented as two different config items.

Reply:
Yeah, this is me wanting not to change the code too much, I think this can be fixed by creating new sessions in the for loop always and then use the session directly in `ecr.New` without additional config.